### PR TITLE
Max profit challenge

### PR DESCRIPTION
**Pull request by misik-eng**
---
name: Best Time to Buy and Sell Stock Challenge
about: A template PR for code review with a checklist
---
**Challenge Description**
Determine the maximum profit that can be achieved from buying and selling a single stock on different days.

**Behavior:**
The function takes an array called prices, where each element prices[i] represents the stock price on day i.
The function returns the maximum profit you can achieve by choosing a single day to buy one stock and a different day after that to sell it. If no profit can be made, return 0.

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_max_profit.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on 12 28 2024gi
+
+@author: Muhammet Isik
+"""
```
Comment (doctorbanu) [inline solutions/tests/test_max_profit.py:7]: It looks like the test module is missing a module-level docstring. Adding one that outlines the purpose of the tests (e.g., covering standard cases, edge cases, and defensive scenarios) would make the module more complete. Also, organizing the test cases under these headings might make them easier to follow.

File: solutions/max_profit.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    >>> maxProfit([3, 3, 3, 3, 3])
+    0  # No profit possible as prices stay the same
+    """
+    # Check input validity, Input must be a integer list
+    assert isinstance(prices, list), "Input must be a list"
+
+    for price in prices:
+        assert isinstance(price, int), "Every single List element must be an integer"
+
+    buy = 0  # Buy index
```
Comment (doctorbanu) [inline solutions/max_profit.py:50]: For the comments on the lines 50 and 51, changing "Buy index" and "Sell index" to "Buy day" and "Sell day" could make the purpose of these variables clearer, as they represent specific days rather than just indices.

File: solutions/max_profit.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+A module for calculating the maximum profit from stock prices.
+
+Module contents:
+    - maxProfit: calculates the maximum profit from a list of stock prices
```
Comment (doctorbanu) [inline solutions/max_profit.py:7]: The file name (max_profit.py) is excellent, but I noticed that “maxProfit” is used in some places of the doctests and docstrings and also the docstring of the test class. It might be clearer to consistently use “max_profit” throughout to align with Python's naming conventions and avoid confusion.

File: solutions/max_profit.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+
+def max_profit(prices: list[int]) -> int:
+    """Calculates the maximum profit from a list of stock prices.
+
+    The function determines the maximum profit that can be achieved by buying
+    and selling a stock once. It ensures that the selling day is after the
+    buying day.
+
+    Parameters:
```
Comment (doctorbanu) [inline solutions/max_profit.py:22]: The function’s docstring could include a little more details about the input. For example, can stock prices be negative? Adding this information would make it easier to understand how the function works.

**Review by doctorbanu — CHANGES_REQUESTED** [review]
Hi Muhammet. I loved the challenge and everything. Great work on this! It’s clear you’ve put in a lot of effort. With a few minor adjustments, it could be even more polished, I think. Looking forward to see the updated version! Thank you in advance.

File: solutions/max_profit.py (hunk 3 above)
Comment (misik-eng) [reply solutions/max_profit.py:7]: Yepp MaxProfit ( Camelcase) is my old function name I forgot to change in doctests, Thx for reminding.. I'll take care of it.

File: solutions/max_profit.py (hunk 4 above)
Comment (misik-eng) [reply solutions/max_profit.py:22]: You're absolutely right, I'll add more detailed explanations about the input, including clarifying that stock prices cannot be negative. Thank you.

File: solutions/max_profit.py (hunk 2 above)
Comment (misik-eng) [reply solutions/max_profit.py:50]: I see your point, and I completely agree. I'll make the change to 'Buy day' and 'Sell day' right away.

Comment (misik-eng) [reply solutions/max_profit.py:50]: I changed variable names to sell_day and buy_day and it is more clear and easy to understand now.

File: solutions/tests/test_max_profit.py (hunk 1 above)
Comment (misik-eng) [reply solutions/tests/test_max_profit.py:7]: Yes I forgot to add explanation in docstring(there is only creation date and author info in module docstring ) , sorry I know it is so absurd, I was working multiple challenges and missed this one. I'll add it right away.

**Review by doctorbanu — APPROVED** [review]

**Comment by doctorbanu** [comment]
Excellent work Muhammet 🙌 Thank you for your efforts.

**Comment by misik-eng** [comment]
There are 5 file changes because I re-added some files that were accidentally deleted (the Factorial challenge function and its test).
I could have done it in another PR .
